File: tests/validation/test_validate.py
"""Tests for the Python validation helper. Run: python3 -m pytest test_validate.py -q"""

import os
import sys

import pytest

sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)),
                                "..", "..", "templates", "project", "scripts"))

from validate import ValidationError, check, init, summary  # noqa: E402


@pytest.fixture()
def log(tmp_path):
    prefix = str(tmp_path / "checks")
    init(prefix, step="test")
    return prefix


def test_exact_pass_writes_readable_line(log):
    assert check("transcripts", 50000, 50000, source="GENCODE v44") is True
    text = open(log + ".log").read()
    assert "PASS" in text
    assert "transcripts: 50,000 (expected 50,000)" in text
    assert "[GENCODE v44]" in text


def test_mismatch_stops_the_run(log):
    with pytest.raises(ValidationError):
        check("transcripts", 49999, 50000)
    assert "FAIL" in open(log + ".log").read()


def test_off_by_one_is_a_failure_not_a_rounding_pass(log):
    with pytest.raises(ValidationError):
        check("samples", 23, 24)


def test_tolerance_band_edges(log):
    assert check("proportion", 0.51, 0.50, tol=0.01)
    with pytest.raises(ValidationError):
        check("proportion", 0.52, 0.50, tol=0.01)


def test_logical_assertion(log):
    assert check("ids preserved", {1, 2} == {1, 2}, True)
    with pytest.raises(ValidationError):
        check("ids preserved", {1, 2} == {1, 3}, True)


def test_zero_actual_is_not_silently_truthy(log):
    with pytest.raises(ValidationError):
        check("cells retained", 0, 5000)


def test_warn_only_records_but_continues(log):
    assert check("optional count", 5, 6, warn_only=True) is False
    assert "WARN" in open(log + ".log").read()


def test_tsv_is_machine_readable(log):
    check("transcripts", 50000, 50000, source="GENCODE v44")
    rows = open(log + ".tsv").read().strip().split("\n")
    assert rows[0].split("\t") == [
        "status", "timestamp", "step", "label", "actual", "expected", "source",
    ]
    fields = rows[1].split("\t")
    assert fields[0] == "PASS"
    assert fields[2] == "test"
    assert fields[3] == "transcripts"
    assert fields[6] == "GENCODE v44"


def test_summary_tally(log):
    check("a", 1, 1)
    check("b", 2, 3, warn_only=True)
    assert summary() == {"pass": 1, "fail": 1}
    assert "2 checks: 1 passed, 1 failed" in open(log + ".log").read()


def test_init_truncates_previous_log(tmp_path):
    prefix = str(tmp_path / "checks")
    init(prefix, step="first")
    check("a", 1, 1)
    init(prefix, step="second")
    assert "first" not in open(prefix + ".log").read()
    assert len(open(prefix + ".tsv").read().strip().split("\n")) == 1

File: tests/validation/test_validate.R
# Tests for the R validation helper. Run: Rscript test_validate.R
# Base R only, so it runs anywhere the analysis containers run.

here <- dirname(normalizePath(sub("^--file=", "",
  grep("^--file=", commandArgs(trailingOnly = FALSE), value = TRUE)[1])))
source(file.path(here, "..", "..", "templates", "project", "scripts", "validate.R"))

failures <- 0L
expect <- function(label, ok) {
  if (!isTRUE(ok)) {
    failures <<- failures + 1L
    cat("TEST FAIL:", label, "\n")
  } else {
    cat("test ok:", label, "\n")
  }
}
fails_with_error <- function(expr) inherits(try(expr, silent = TRUE), "try-error")

tmp <- file.path(tempdir(), "vtest")
dir.create(tmp, showWarnings = FALSE)
prefix <- file.path(tmp, "checks")

# --- exact pass, readable line ------------------------------------------------
vcheck_init(prefix, step = "test")
vcheck("transcripts", 50000, 50000, source = "GENCODE v44")
text <- paste(readLines(paste0(prefix, ".log")), collapse = "\n")
expect("pass is logged", grepl("PASS", text))
expect("count is thousands-separated", grepl("transcripts: 50,000 \\(expected 50,000\\)", text))
expect("source is recorded", grepl("\\[GENCODE v44\\]", text))

# --- a mismatch stops the run -------------------------------------------------
expect("off-by-one fails", fails_with_error(vcheck("transcripts", 49999, 50000)))
expect("zero is not silently truthy", fails_with_error(vcheck("cells", 0, 5000)))

# --- tolerance band edges -----------------------------------------------------
vcheck_init(prefix, step = "test")
expect("inside tolerance passes", isTRUE(vcheck("proportion", 0.51, 0.50, tol = 0.01)))
expect("outside tolerance fails", fails_with_error(vcheck("proportion", 0.52, 0.50, tol = 0.01)))

# --- logical assertions -------------------------------------------------------
vcheck_init(prefix, step = "test")
expect("set equality passes", isTRUE(vcheck("ids preserved", setequal(1:2, 2:1), TRUE)))
expect("set inequality fails", fails_with_error(vcheck("ids preserved", setequal(1:2, c(1, 3)), TRUE)))
expect("NA actual fails", fails_with_error(vcheck("count", NA_integer_, 10)))

# --- warn_only records but continues -----------------------------------------
vcheck_init(prefix, step = "test")
expect("warn_only returns FALSE", identical(vcheck("optional", 5, 6, warn_only = TRUE), FALSE))
expect("warn is logged", any(grepl("^WARN", readLines(paste0(prefix, ".log")))))

# --- machine-readable tsv -----------------------------------------------------
vcheck_init(prefix, step = "01_load")
vcheck("transcripts", 50000, 50000, source = "GENCODE v44")
rows <- readLines(paste0(prefix, ".tsv"))
expect("tsv header", identical(strsplit(rows[1], "\t")[[1]],
  c("status", "timestamp", "step", "label", "actual", "expected", "source")))
f <- strsplit(rows[2], "\t")[[1]]
expect("tsv row fields", f[1] == "PASS" && f[3] == "01_load" && f[4] == "transcripts" &&
  f[7] == "GENCODE v44")

# --- summary tally ------------------------------------------------------------
vcheck_init(prefix, step = "test")
vcheck("a", 1, 1)
vcheck("b", 2, 3, warn_only = TRUE)
tally <- vcheck_summary()
expect("tally counts", identical(tally, list(pass = 1L, fail = 1L)))
expect("tally line", any(grepl("2 checks: 1 passed, 1 failed", readLines(paste0(prefix, ".log")))))

# --- init truncates a previous log -------------------------------------------
vcheck_init(prefix, step = "first")
vcheck("a", 1, 1)
vcheck_init(prefix, step = "second")
expect("log truncated", !grepl("first", paste(readLines(paste0(prefix, ".log")), collapse = "")))
expect("tsv truncated", length(readLines(paste0(prefix, ".tsv"))) == 1L)

cat(if (failures == 0L) "\nall R validation tests passed\n" else
    sprintf("\n%d R validation tests FAILED\n", failures))
quit(status = if (failures == 0L) 0L else 1L)
